Statically check that number of placeholders in log message matches number of parameters

ywelsch opened the pull request:
This PR checks that the number of placeholders in log message string matches number of parameters passed to logging call.

Good:

```
logger.warn("Hello {}", name)
logger.warn("Hello {}", throwable, name)
```

Bad:

```
logger.warn("Hello {}", firstName, lastName)
logger.warn("Hello {}", name, throwable)
```

Checks are run using the new Gradle task `loggerUsageCheck`.

To fully check all occurrences, the checker requires that log messages are constant strings. In case where a logging usage should be ignored, an annotation `@SuppressLoggerChecks` can be put on the respective method or class where logging occurrences are to be ignored.

ywelsch committed:
Check that number of placeholders in log message matches number of parameters

ywelsch committed:
Fix wrong placeholder usage in logging statements

dakrone commented:
Hooray for this feature, thanks @ywelsch!

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/LoggerUsageTask.groovy:
```diff
@@ -0,0 +1,76 @@
+                }
+            }
+            doFirst({
+                args('-cp', classpath.asPath, 'org.elasticsearch.qa.loggerusage.ESLoggerUsageChecker')
```
It feels weird having the check live in qa. I believe it is more gradle-ish to have the check's source code live in buildSrc. And be written in groovy. Though we write very java-like groovy.

The idea, I believe, is that we can extract bits of buildSrc and make them into published gradle plugins.

nik9000 commented on qa/logger-usage/build.gradle:
```diff
@@ -0,0 +1,43 @@
+}
+
+// TODO: should we have licenses for our test deps?
+dependencyLicenses.enabled = false
```
If it were in buildSrc I don't think it'd have to do this. OTOH it should probably follow the same rules and declare the dependency.

nik9000 commented on qa/logger-usage/src/main/java/org/elasticsearch/qa/loggerusage/ESLoggerUsageChecker.java:
```diff
@@ -0,0 +1,437 @@
+    public static final List<String> LOGGER_METHODS = Arrays.asList("trace", "debug", "info", "warn", "error");
+
+    public static void main(String... args) throws Exception {
+        System.out.println("checking for wrong usages of ESLogger...");
```
Yeah, this should be in buildSrc and it should use a ProgressLogger.

nik9000 commented on qa/logger-usage/src/main/java/org/elasticsearch/qa/loggerusage/ESLoggerUsageChecker.java:
```diff
@@ -0,0 +1,437 @@
+        throws IOException {
+        for (String classDirectory : classDirectories) {
+            File dir = new File(classDirectory);
+            if (dir.isDirectory()) {
```
Something like:
```groovy
if (false == (dir.exists() && dir.isDirectory())) {
  throw new InvalidUserDataException(dir + ' should be an existing directory.');
}
```

As in, should we just fail if the user points it as some garbage directory?

nik9000 commented on qa/logger-usage/src/main/java/org/elasticsearch/qa/loggerusage/ESLoggerUsageChecker.java:
```diff
@@ -0,0 +1,437 @@
+        }
+
+        /**
+         * Returns an error message that has the form of stack traces emitted by {@link Throwable#printStackTrace}
```
form of *a single line from* {@link Throwable#printStackTrace}?

nik9000 commented on qa/logger-usage/src/main/java/org/elasticsearch/qa/loggerusage/ESLoggerUsageChecker.java:
```diff
@@ -0,0 +1,437 @@
+
+        @Override
+        public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
+            if (methodName == null || methodName.equals(name)) {
```
It **might** be more clear to replace methodName with `Predicate<String> methodsToCheck`. It would have helped me when I first scanned the code.

nik9000 commented on qa/logger-usage/src/main/java/org/elasticsearch/qa/loggerusage/ESLoggerUsageChecker.java:
```diff
@@ -0,0 +1,437 @@
+                    BasicValue logMessageLengthObject = getStackValue(stringFrames[i], argumentTypes.length - 1); // first argument
+                    if (logMessageLengthObject instanceof PlaceHolderStringBasicValue == false) {
+                        // we should probably require that only string constants are passed as first argument
+                        // badUsages.add(new WrongLoggerUsage(className, methodNode.name, lineNumber,
```
Yeah. If we don't then the check just silently skips the invocation. Something like "I don't understand log invocations of this type."

nik9000 commented:
I talked to @ywelsch about these in person because we had both flown to San Francisco to Elastic{ON}:

1. The check is its own project so it can depend on test-framework and elasticsearch core which makes it much much easier to implement.
2. Embedding it into the gradle process forces you to use their version of byte code library who's version doesn't match. Or you could shade it. Either way that complexity isn't worth it when you can just fork.
3. It does indeed feel weird to have it in `qa`. Maybe move it to the `test` directory or something? Once you accept that it should be a subproject then exactly where you put it is simpler change after the fact so its ok to pay less attention to during the PR.


I'm super excited to get this in though!

ywelsch committed:
Make logging message String constant to allow static checks

ywelsch commented:
@nik9000 I've updated the PR with suggested changes (last 3 commits). I've also modified all logger calls where the log message was not a constant string. This ensures that ALL logger calls are correct. In case where people want to opt-out of the checking mechanism (e.g. `NettyInternalESLogger` which delegates to `ESLogger`), I've added the `@SuppressLoggerChecks` annotation.

Note that log output is of the following form (each error being represented by two lines):
```
Bad usage of org.elasticsearch.common.logging.ESLogger#debug: Expected 4 arguments but got 5
	org.elasticsearch.action.ingest.IngestActionFilter.lambda$processBulkIndexRequest$198(IngestActionFilter.java:106)
Bad usage of org.elasticsearch.common.logging.ESLogger#trace: Expected 1 arguments but got 0
	org.elasticsearch.cluster.action.index.NodeIndexDeletedAction.nodeIndexDeleted(NodeIndexDeletedAction.java:84)
Bad usage of org.elasticsearch.common.logging.ESLogger#trace: Expected 2 arguments but got 3
	org.elasticsearch.cluster.action.shard.ShardStateAction$2.onNewClusterState(ShardStateAction.java:154)
Bad usage of org.elasticsearch.common.logging.ESLogger#debug: Expected 0 arguments but got 1
	org.elasticsearch.common.geo.builders.PolygonBuilder.component(PolygonBuilder.java:317)
Bad usage of org.elasticsearch.common.logging.ESLogger#debug: Expected 0 arguments but got 1
	org.elasticsearch.common.geo.builders.PolygonBuilder.buildCoordinates(PolygonBuilder.java:395)
Bad usage of org.elasticsearch.common.logging.ESLogger#warn: Expected 2 arguments but got 3
	org.elasticsearch.common.lucene.Lucene.parseVersion(Lucene.java:114)
```

rjernst commented:
I don't think we should do this as a project under test. It is ultimately a gradle task. If we need to separate the asm version, it can still be a separate subproject within buildSrc.

ywelsch commented:
@rjernst I also want to have real tests for the checker (see `ESLoggerUsageTests`). This means:
- I need access to the `core` project from the tests
- I want to make use of the testing infrastructure provided by `test:framework`

How would I do that while having the sources of the checker under `buildSrc`?
Also, there is no real infrastructure under `buildSrc` right now to run tests...

You can compare this checker a bit to the JarHell task, which also lives under core...

rjernst commented:
JarHell is something we use in core at runtime, which is exposed with a main method to allow using from the command line like we do before tests. But this here is strictly a static check for builds.

Adding a subproject in buildSrc is easy. And I don't think we should be using the test framework for testing it, since that is a huge overkill for testing whether this simple utility works. You can set up tests that use the normal gradle test runner with junit, and in the test src add ESLogger so the tests can work (ie find the ESLogger class, just with a mock of it).

nik9000 commented:
> Adding a subproject in buildSrc is easy

I'll have a look at it this afternoon-ish and see if I can send a PR to ywelsch:enhance/elasticsearch-proper-logging-usage that does that.

ywelsch commented:
@rjernst The issue with subprojects of buildSrc is that they are not well supported:
- They do not properly integrate into IDE's. IntelliJ is unable to detect buildSrc subprojects, requiring a separate project import of buildSrc as top-level project to work on the buildSrc subprojects.
- They are not added to the buildscript classpath, requiring the actual Gradle task to be in buildSrc (or some hacks as workaround, see http://stackoverflow.com/questions/23990513/how-to-import-multiproject-build-in-buildsrc-dir-in-gradle)

Also:
- If one of the tests were to fail, the project cannot be imported into / refreshed in IDEA. Any gradle command would fail (as `buildSrc:build` is run first). This behavior would be ok if one of the core parts of the build infrastructure would be broken, but probably not for a helper tool that checks logging lines.
- Making the tests work requires mocking ESLogger in the test src. This requires essentially duplicating ESLogger class and introducing a check somewhere that that we are compatible with the actual ESLogger in core.

The current approach does not have to deal with these issues, is simple and integrates nicely with the build. Another advantage of the current approach is that the classpath of the tool is clean (not polluted with gradle api).

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy:
```diff
@@ -109,4 +110,22 @@ class PrecommitTasks {
+    private static Task configureLoggerUsage(Project project) {
+        Task loggerUsageTask = project.tasks.create('loggerUsageCheck', LoggerUsageTask.class)
+
+        project.configurations {
```
We've been moving towards a more java like style in buildSrc. Can you use the direct methods to create configurations/dependencies, instead of closures?

ywelsch replied:
ok, addressed in cbec014

rjernst commented on test/logger-usage/build.gradle:
```diff
@@ -0,0 +1,27 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
```
Can this use the `elasticsearch.build` plugin? I noticed the code is using eg File instead of Path, which would have been caught by forbidden apis. I realize it is sort of a chicken-and-egg problem, but I think we should be able to compose this in such a way that the task which depends on this project is not added? Perhaps it is worth a hardcoded conditional on the project path in PrecommitTasks.

ywelsch replied:
it is using `elasticsearch.build` already as that is applied through parent project (`test`)

rjernst replied:
Ok, I see the problem now...forbidden apis and jarhell are disabled in `test/build.gradle` for subprojects...we really need to re-enable that...separate issue I guess

rjernst commented:
Thanks @ywelsch, I did not realize the problems intellij has with buildSrc containing subprojects. I left a couple thoughts about the gradle side. I didn't look through the actual checking code that closely, but the tests look pretty good so I am happy enough there for now.

ywelsch commented:
@rjernst I've pushed a commit that addresses your comments and enables forbiddenApis for the `logger-usage` project. As this project does not rely on `core`, I had to split the forbiddenApis signatures:

- all-signatures is split into jdk-signatures and es-all-signatures. The reason this is needed is that forbiddenApis tries to load the classes that are in the signatures file (which fails for the logging-usage project as it does not rely on es/core)
- For consistency, I then renamed core-signatures to es-core-signatures and test-signatures to es-test-signatures.

If you don't like the splitting, we can simply simply disable forbiddenApis for `logger-usage`.

rjernst commented:
Thanks @ywelsch. LGTM.

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/LoggerUsageTask.groovy:
```diff
@@ -0,0 +1,76 @@
+
+    /** The classpath to run logger usage on */
+    @InputFile
+    public FileCollection classpath;
```
I **think** putting public here breaks gradle's detection of "up to date"-ness. It has something to do with the difference between public fields (what happens when you declare a field public) and groovy properties (what happens when you leave off the access modifier). I can't find the stackoverflow post where someone talked about it, but see if up-to-date checking works for your task. If not try ripping off the `public` and trying again.

ywelsch replied:
Good catch!
The following resource explains it: https://groups.google.com/forum/#!topic/gradle-dev/EC2bhFUqDzg
I've converted to private fields with explicit getters. JarHell task (and possibly others) has the same issue (will make a follow-up commit for that)

nik9000 replied:
> has the same issue

Yeah. I noticed that and then forgot to send PRs for it.

nik9000 commented on core/src/main/java/org/elasticsearch/bootstrap/JVMCheck.java:
```diff
@@ -118,7 +118,7 @@ static void check() {
             if (bug != null) {
                 if (bug.workAround != null && ManagementFactory.getRuntimeMXBean().getInputArguments().contains(bug.workAround)) {
-                    Loggers.getLogger(JVMCheck.class).warn(bug.getWarningMessage());
+                    Loggers.getLogger(JVMCheck.class).warn("{}", bug.getWarningMessage());
```
The old way should have been ok though?

nik9000 replied:
Ah. I get it. You are checking that the first string is always a constant. Sometimes it makes for weird things like this but that should be ok.

ywelsch replied:
yeah, there are a few cases where you have to put "{}", but the advantage is that toString() is only evaluated then when we log at that level. 
Related: http://www.slf4j.org/faq.html#string_contents

nik9000 commented on modules/lang-groovy/src/test/java/org/elasticsearch/messy/tests/ExtendedStatsTests.java:
```diff
@@ -503,7 +503,7 @@ private void assertShardExecutionState(SearchResponse response, int expectedFail
         if (failures.length != expectedFailures) {
             for (ShardSearchFailure failure : failures) {
-                logger.error("Shard Failure: {}", failure.reason(), failure.toString());
+                logger.error("Shard Failure: {}", failure);
```
Maybe the original author meant `logger.error("Shard Failure: {}", failure.cause(), failure);`?

nik9000 commented on modules/lang-groovy/src/test/java/org/elasticsearch/messy/tests/StatsTests.java:
```diff
@@ -404,10 +404,10 @@ private void assertShardExecutionState(SearchResponse response, int expectedFail
         if (failures.length != expectedFailures) {
             for (ShardSearchFailure failure : failures) {
-                logger.error("Shard Failure: {}", failure.reason(), failure.toString());
+                logger.error("Shard Failure: {}", failure);
```
Same here.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/CorruptionUtils.java:
```diff
@@ -88,11 +88,11 @@ public static void corruptFile(Random random, Path... files) throws IOException
             msg.append(" checksum value after corruption: ").append(actualChecksumAfterCorruption).append("]");
             msg.append(" file: ").append(fileToCorrupt.getFileName()).append(" length: ").append(dir.fileLength(fileToCorrupt.getFileName().toString()));
-            logger.info(msg.toString());
+            logger.info("Checksum info: {}", msg);
```
You've changed the format of this slightly. Probably best to just remove `info: `.

nik9000 commented on test/logger-usage/src/main/java/org/elasticsearch/test/loggerusage/ESLoggerUsageChecker.java:
```diff
@@ -0,0 +1,460 @@
+                    BasicValue logMessageLengthObject = getStackValue(stringFrames[i], argumentTypes.length - 1); // first argument
+                    if (logMessageLengthObject instanceof PlaceHolderStringBasicValue == false) {
+                        wrongUsageCallback.accept(new WrongLoggerUsage(className, methodNode.name, methodInsn.name, lineNumber,
+                            "First argument must be a string constant"));
```
Maybe add "so that we can statically check it for errors"?

nik9000 commented:
I left a few small notes. I don't know ASM so I'm going to trust you there but I suspect it is fine. It found a lot of things! My vote is to merge it as soon as you can resolve the conflicts so it starts catching errors.

ywelsch commented:
@nik9000 Addressed all your comments. I will go ahead and push.
Thanks @nik9000 and @rjernst for the reviews!

ywelsch commented:
I've cherry-picked wrong usage of placeholders in log statements (718876a) to 2.x (b24d6b3). I've NOT cherry-picked all the logging changes that were made to make log messages string constants.
